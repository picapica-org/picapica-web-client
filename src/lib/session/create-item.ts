import { CreateItemRequest } from "../generated/v1/services_pb";
import { Item } from "../generated/v1/types_pb";
import { assertNever, shorten, shortenWords } from "../util";

export type ItemData =
	| { readonly type: "file"; file: File }
	| { readonly type: "url"; url: string }
	| { readonly type: "text"; text: string };

export type ItemType = ItemData["type"];

/**
 * An item prototype.
 *
 * This simply contains the name of an item and from where to get its data.
 */
export class ItemProto {
	readonly name: string;
	readonly data: ItemData;

	constructor(name: string, data: ItemData) {
		this.name = name;
		this.data = data;
	}

	static fromFile(file: File): ItemProto {
		return new ItemProto(file.name, { type: "file", file });
	}
	static fromUrl(url: string): ItemProto {
		return new ItemProto(shorten(url, 100), { type: "url", url });
	}
	static fromText(text: string): ItemProto {
		return new ItemProto(extractNameFromText(text), { type: "text", text });
	}

	getMeta(): ItemMeta {
		return {
			name: this.name,
			type: this.data.type,
			size: getByteSize(this.data),
		};
	}

	async read(): Promise<ItemRawProto> {
		const raw = await readData(this.data);
		return new ItemRawProto(this.name, this.data.type, raw);
	}
}

/**
 * An item prototype with its raw data read into memory.
 */
export class ItemRawProto {
	readonly name: string;
	readonly type: ItemType;
	readonly raw: Uint8Array;

	constructor(name: string, type: ItemType, raw: Uint8Array) {
		this.name = name;
		this.type = type;
		this.raw = raw;
	}

	getMeta(): ItemMeta {
		return {
			name: this.name,
			type: this.type,
			size: this.raw.byteLength,
		};
	}

	getRequest(sessionUrn: string): CreateItemRequest {
		const req = new CreateItemRequest();

		req.setSessionUrn(sessionUrn);
		req.setMeta(new Item.Metadata().setName(this.name));
		req.setType(toItemResourceType(this.type));
		req.setRaw(this.raw);

		return req;
	}
}

export interface ItemMeta {
	readonly name: string;
	readonly type: ItemType;
	readonly size: number;
}

const encoder = new TextEncoder();
async function readData(data: ItemData): Promise<Uint8Array> {
	switch (data.type) {
		case "file":
			return new Uint8Array(await new Response(data.file).arrayBuffer());
		case "url":
			return encoder.encode(data.url);
		case "text":
			return encoder.encode(data.text);
		default:
			assertNever(data);
	}
}

function getByteSize(data: ItemData): number {
	switch (data.type) {
		case "file":
			return data.file.size;
		case "url":
			return encoder.encode(data.url).byteLength;
		case "text":
			return encoder.encode(data.text).byteLength;
		default:
			assertNever(data);
	}
}

export function toItemResourceType(type: ItemType): Item.Resource.Type {
	switch (type) {
		case "file":
			return Item.Resource.Type.TYPE_FILE;
		case "url":
			return Item.Resource.Type.TYPE_URL;
		case "text":
			return Item.Resource.Type.TYPE_TEXT;
		default:
			assertNever(type);
	}
}
export function toItemType(type: Item.Resource.Type): ItemType | undefined {
	switch (type) {
		case Item.Resource.Type.TYPE_UNSPECIFIED:
			return undefined;
		case Item.Resource.Type.TYPE_FILE:
			return "file";
		case Item.Resource.Type.TYPE_URL:
			return "url";
		case Item.Resource.Type.TYPE_TEXT:
			return "text";
		default:
			assertNever(type);
	}
}

function extractNameFromText(text: string): string {
	return shortenWords(text.replace(/\s+/g, " ").trim(), 30, 50);
}
