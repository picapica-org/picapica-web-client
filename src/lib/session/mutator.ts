import { Session } from "../generated/v1/services_pb";
import { DeepReadonly, identity } from "../util";

export type SessionMutator = (session: DeepReadonly<Session.AsObject>) => DeepReadonly<Session.AsObject>;

export const noMutation: SessionMutator = identity;
