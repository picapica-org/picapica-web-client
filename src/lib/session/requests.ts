import { CreateItemRequest } from "../generated/v1/services_pb";
import { Item } from "../generated/v1/types_pb";
import { shorten } from "../util";

const encoder = new TextEncoder();

export function newCreateTextItemRequest(text: string): CreateItemRequest {
	const req = new CreateItemRequest();

	req.setType(Item.Resource.Type.TYPE_TEXT);
	req.setRaw(encoder.encode(text));

	// Set the name to the first line
	const name = /^.*/.exec(text.trim())![0];
	req.setMeta(new Item.Metadata().setName(shorten(name, 100)));

	return req;
}
export function newCreateUrlItemRequest(url: string): CreateItemRequest {
	const req = new CreateItemRequest();

	req.setType(Item.Resource.Type.TYPE_URL);
	req.setRaw(encoder.encode(url));
	req.setMeta(new Item.Metadata().setName(shorten(url, 100)));

	return req;
}
