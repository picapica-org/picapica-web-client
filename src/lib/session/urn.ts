import { parse as parseURI, URIComponents } from "uri-js";

export const NONE_URN = "urn:none:";

export type PicapicaUrn =
	| PicapicaCollectionUrn
	| PicapicaDocumentUrn
	| PicapicaSessionUrn
	| PicapicaItemUrn
	| PicapicaResultUrn;
export interface PicapicaCollectionUrn {
	type: "collection";
	collectionId: string;
}
export interface PicapicaDocumentUrn {
	type: "document";
	collectionId: string;
	documentId: string;
}
export interface PicapicaSessionUrn {
	type: "session";
	sessionId: string;
}
export interface PicapicaItemUrn {
	type: "item";
	sessionId: string;
	itemId: string;
}
export interface PicapicaResultUrn {
	type: "result";
	sessionId: string;
	resultId: string;
}

// eslint-disable-next-line @typescript-eslint/no-namespace, @typescript-eslint/no-redeclare
export namespace PicapicaUrn {
	export function parse(urn: string): PicapicaUrn {
		const parsed = parseURI(urn) as URIComponents & { nid?: string; nss?: string };
		if (parsed.scheme !== "urn" || parsed.nid !== "picapica" || !parsed.nss) {
			throw new SyntaxError(`Invalid Picapica URN: ${urn}`);
		}

		const parts = parsed.nss.split(/:/);
		const type = parts[0];
		const args = parts.slice(1);

		switch (type) {
			case "collection":
				if (args.length === 1) {
					return { type, collectionId: args[0] };
				}
				throw new SyntaxError(`Invalid Picapica ${type} URN: ${urn}`);

			case "document":
				if (args.length === 2) {
					return { type, collectionId: args[0], documentId: args[1] };
				}
				throw new SyntaxError(`Invalid Picapica ${type} URN: ${urn}`);

			case "session":
				if (args.length === 1) {
					return { type, sessionId: args[0] };
				}
				throw new SyntaxError(`Invalid Picapica ${type} URN: ${urn}`);

			case "item":
				if (args.length === 2) {
					return { type, sessionId: args[0], itemId: args[1] };
				}
				throw new SyntaxError(`Invalid Picapica ${type} URN: ${urn}`);

			case "result":
				if (args.length === 2) {
					return { type, sessionId: args[0], resultId: args[1] };
				}
				throw new SyntaxError(`Invalid Picapica ${type} URN: ${urn}`);

			default:
				throw new SyntaxError(`Invalid Picapica URN: ${urn}`);
		}
	}

	export function stringify(urn: PicapicaUrn): string {
		return `urn:picapica:${urn.type}${stringifyContent(urn)}`;
	}
	function stringifyContent(urn: PicapicaUrn): string {
		switch (urn.type) {
			case "collection":
				return `:${urn.collectionId}`;
			case "document":
				return `:${urn.collectionId}:${urn.documentId}`;
			case "session":
				return `:${urn.sessionId}`;
			case "item":
				return `:${urn.sessionId}:${urn.itemId}`;
			case "result":
				return `:${urn.sessionId}:${urn.resultId}`;
		}
	}
}
