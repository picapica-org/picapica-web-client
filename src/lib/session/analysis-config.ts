import { ResourcePair } from "../generated/v1/types_pb";
import { EMPTY_MAP, EMPTY_SET } from "../util";
import { NONE_URN, PicapicaUrn } from "./urn";

export type ItemUrn = string & { readonly __itemUrn?: never };
export type CollectionUrn = string & { readonly __itemUrn?: never };

export class AnalysisConfig {
	readonly groupA: ReadonlySet<ItemUrn>;
	readonly groupB: ReadonlySet<ItemUrn>;
	readonly collections: ReadonlyMap<CollectionUrn, ReadonlySet<ItemUrn>>;

	constructor(
		groupA: ReadonlySet<ItemUrn>,
		groupB: ReadonlySet<ItemUrn>,
		collections: ReadonlyMap<CollectionUrn, ReadonlySet<ItemUrn>>
	) {
		this.groupA = groupA;
		this.groupB = groupB;
		this.collections = collections;
	}

	withGroupA(groupA: ReadonlySet<ItemUrn>): AnalysisConfig {
		return new AnalysisConfig(groupA, this.groupB, this.collections);
	}
	withGroupB(groupB: ReadonlySet<ItemUrn>): AnalysisConfig {
		return new AnalysisConfig(this.groupA, groupB, this.collections);
	}
	withCollections(collections: ReadonlyMap<CollectionUrn, ReadonlySet<ItemUrn>>): AnalysisConfig {
		return new AnalysisConfig(this.groupA, this.groupB, collections);
	}
	withCollection(collection: CollectionUrn, items: ReadonlySet<ItemUrn>): AnalysisConfig {
		const copy = new Map(this.collections);
		copy.set(collection, items);
		return this.withCollections(copy);
	}

	/**
	 * Creates a new analysis config from the given collection of resource pairs.
	 *
	 * The approach implemented to compress O(n^2) item combinations into 2 O(n) groups is very
	 * simplistic and works as follows: If a resource pair contains 2 item URNs, then the `a` URN
	 * will be in group A and the `b` URN will be in group B.
	 *
	 * Disappointingly simple, right? Well, it does have a few advantages: It's fast and it doesn't
	 * lose any combination as a byproduct of the lossy conversion. However, it might cause some
	 * combinations to be added.
	 *
	 * This approach also has the very nice property that:
	 * `a == fromResourcePairs(a.getResourcePairs())`.
	 *
	 * @param pairs
	 */
	static fromResourcePairs(
		pairs: Iterable<ResourcePair | ResourcePair.AsObject>,
		allItems: ReadonlyArray<ItemUrn>
	): AnalysisConfig {
		const groupA = new Set<ItemUrn>();
		const groupB = new Set<ItemUrn>();
		const collections = new Map<CollectionUrn, Set<ItemUrn>>();

		function addCollection(collection: CollectionUrn, item: ItemUrn): void {
			let set = collections.get(collection);
			if (set === undefined) {
				set = new Set();
				collections.set(collection, set);
			}
			set.add(item);
		}
		function toItemUrns(urn: string): ReadonlyArray<ItemUrn> {
			const type = getPicapicaUrnType(urn);

			if (type === "item") {
				return [urn];
			} else if (type === "session") {
				return allItems;
			} else {
				return [];
			}
		}

		for (const pair of pairs) {
			const [a, b] = pair instanceof ResourcePair ? [pair.getUrnA(), pair.getUrnB()] : [pair.urnA, pair.urnB];

			const aType = getPicapicaUrnType(a);
			const bType = getPicapicaUrnType(b);

			if (aType === "collection") {
				toItemUrns(b).forEach(item => addCollection(a, item));
				continue;
			}
			if (bType === "collection") {
				toItemUrns(a).forEach(item => addCollection(b, item));
				continue;
			}

			toItemUrns(a).forEach(item => groupA.add(item));
			toItemUrns(b).forEach(item => groupB.add(item));
		}

		return new AnalysisConfig(groupA, groupB, collections);
	}

	/**
	 * Returns a list of resource pairs that represents this config.
	 *
	 * @returns
	 */
	getResourcePairs(): ResourcePair[] {
		const pairs: ResourcePair[] = [];

		const pairsA = new Set<ItemUrn>();
		const pairsB = new Set<ItemUrn>();

		// add group combinations
		for (const a of this.groupA) {
			for (const b of this.groupB) {
				if (a === b) {
					// same item
				} else {
					pairsA.add(a);
					pairsB.add(b);
					pairs.push(new ResourcePair().setUrnA(a).setUrnB(b));
				}
			}
		}

		// add items from groups that aren't in any pair
		for (const a of this.groupA) {
			if (!pairsA.has(a)) {
				pairs.push(new ResourcePair().setUrnA(a).setUrnB(NONE_URN));
			}
		}
		for (const b of this.groupB) {
			if (!pairsB.has(b)) {
				pairs.push(new ResourcePair().setUrnA(NONE_URN).setUrnB(b));
			}
		}

		// add collections
		for (const [collection, items] of this.collections) {
			for (const item of items) {
				pairs.push(new ResourcePair().setUrnA(collection).setUrnB(item));
			}
		}

		return pairs;
	}
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace AnalysisConfig {
	export const EMPTY = new AnalysisConfig(EMPTY_SET, EMPTY_SET, EMPTY_MAP);
}

function getPicapicaUrnType(urn: string): PicapicaUrn["type"] | undefined {
	try {
		return PicapicaUrn.parse(urn).type;
	} catch (error) {
		return undefined;
	}
}
