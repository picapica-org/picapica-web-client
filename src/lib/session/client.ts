import { SessionServiceClient } from "../generated/v1/ServicesServiceClientPb";
import * as v1_services_pb from "../generated/v1/services_pb";
import { Item } from "../generated/v1/types_pb";
import type * as grpcWeb from "grpc-web";
import { v4 as uuidv4 } from "uuid";
import { delay, lazy } from "../util";
import { PicapicaSessionUrn, PicapicaUrn } from "./urn";

const CLIENT: SessionServiceClient = new SessionServiceClient("http://localhost:8080");
const mock = false;

export const getSessionClient = lazy(() => {
	if (mock) {
		return new MockClient();
	}
	return CLIENT;
});

// Mock client

interface MockUnaryCall<Req, Res> {
	(request: Req, metadata: grpcWeb.Metadata | null): Promise<Res>;
	(
		request: Req,
		metadata: grpcWeb.Metadata | null,
		callback: (err: grpcWeb.Error, response: Res) => void
	): grpcWeb.ClientReadableStream<Res>;
}
function mockUnary<Req, Res>(
	call: (request: Req, metadata: grpcWeb.Metadata | null) => Promise<Res>
): MockUnaryCall<Req, Res> {
	return function getSession(
		request: Req,
		metadata: grpcWeb.Metadata | null,
		callback?: (err: grpcWeb.Error, response: Res) => void
	): any {
		if (callback !== undefined) {
			throw new Error("Streaming calls are not supported.");
		}
		return call(request, metadata);
	};
}

class MockClient extends SessionServiceClient {
	private _sessions = new Map<string, v1_services_pb.Session>();
	private _content = new WeakMap<Item.Resource, string>();

	constructor() {
		super("foo.test");
	}

	private async simulateNetwork(): Promise<void> {
		const DELAY_MIN = 500;
		const DELAY_MAX = 1000;

		await delay(DELAY_MIN + Math.random() * (DELAY_MAX - DELAY_MIN));

		const FAILURE_RATE = 0;

		if (Math.random() < FAILURE_RATE) {
			throw new Error("Network error");
		}
	}

	private async _getSessionRef(urn: string): Promise<v1_services_pb.Session> {
		const session = this._sessions.get(urn);
		if (session === undefined) {
			return Promise.reject("Invalid session id");
		}
		return Promise.resolve(session);
	}
	private async _getItemRef(sessionUrn: string, itemId: string): Promise<Item> {
		const sessionRef = await this._getSessionRef(sessionUrn);

		const item = sessionRef.getItemsList().find(i => i.getUrn() === itemId);
		if (item === undefined) {
			return Promise.reject("Invalid item id");
		}
		return item;
	}

	createSession = mockUnary<v1_services_pb.CreateSessionRequest, v1_services_pb.CreateSessionResponse>(async req => {
		await this.simulateNetwork();

		const urn = PicapicaUrn.stringify({ type: "session", sessionId: uuidv4() });
		const session = new v1_services_pb.Session();
		session.setUrn(urn);
		this._sessions.set(urn, session);

		return new v1_services_pb.CreateSessionResponse().setSessionUrn(urn);
	});
	getSession = mockUnary<v1_services_pb.GetSessionRequest, v1_services_pb.GetSessionResponse>(async req => {
		await this.simulateNetwork();

		const sessionRef = await this._getSessionRef(req.getSessionUrn());
		return new v1_services_pb.GetSessionResponse().setSession(sessionRef.clone());
	});
	deleteSession = mockUnary<v1_services_pb.DeleteSessionRequest, v1_services_pb.DeleteSessionResponse>(async req => {
		await this.simulateNetwork();

		const success = this._sessions.delete(req.getSessionUrn());
		if (!success) {
			return Promise.reject("Invalid session id");
		}
		return new v1_services_pb.DeleteSessionResponse();
	});

	getCollections = mockUnary<v1_services_pb.GetCollectionsRequest, v1_services_pb.GetCollectionsResponse>(
		async req => {
			await this.simulateNetwork();

			return new v1_services_pb.GetCollectionsResponse();
		}
	);

	getConfig = mockUnary<v1_services_pb.GetConfigRequest, v1_services_pb.GetConfigResponse>(async req => {
		await this.simulateNetwork();

		const sessionRef = await this._getSessionRef(req.getSessionUrn());
		return new v1_services_pb.GetConfigResponse().setConfig(sessionRef.getConfig()?.clone());
	});
	updateConfig = mockUnary<v1_services_pb.UpdateConfigRequest, v1_services_pb.UpdateConfigResponse>(async req => {
		await this.simulateNetwork();

		const sessionRef = await this._getSessionRef(req.getSessionUrn());
		sessionRef.setConfig(req.getConfig()?.clone());
		return new v1_services_pb.UpdateConfigResponse();
	});
	deleteConfig = mockUnary<v1_services_pb.DeleteConfigRequest, v1_services_pb.DeleteConfigResponse>(async req => {
		await this.simulateNetwork();

		const sessionRef = await this._getSessionRef(req.getSessionUrn());
		sessionRef.setConfig(undefined);
		return new v1_services_pb.DeleteConfigResponse();
	});

	createItem = mockUnary<v1_services_pb.CreateItemRequest, v1_services_pb.CreateItemResponse>(async req => {
		await this.simulateNetwork();

		const sessionRef = await this._getSessionRef(req.getSessionUrn());

		const raw = req.getRaw_asU8();

		// delay to simulate the upload (100 kB/s)
		await delay(raw.byteLength / 100);

		let content: string;
		switch (req.getType()) {
			case Item.Resource.Type.TYPE_UNSPECIFIED:
				throw new Error("Unspecified resource type");
			case Item.Resource.Type.TYPE_TEXT:
				content = new TextDecoder("utf-8").decode(raw);
				break;
			case Item.Resource.Type.TYPE_URL:
				content = "Text from website: " + new TextDecoder("utf-8").decode(raw);
				break;
			case Item.Resource.Type.TYPE_FILE:
				content = "Some file";
				break;
			default:
				throw new Error("Invalid resource type");
		}

		const itemUrn = PicapicaUrn.stringify({
			type: "item",
			sessionId: (PicapicaUrn.parse(sessionRef.getUrn()) as PicapicaSessionUrn).sessionId,
			itemId: randomHex(64),
		});

		const rawProps = new Item.Resource.RawProperties();
		rawProps.setSize(raw.length);
		rawProps.setChecksum(randomHex(32));

		const processedProps = new Item.Resource.ProcessedProperties();
		processedProps.setLength([...content].length);
		processedProps.setChecksum(randomHex(32));

		const resource = new Item.Resource();
		resource.setItemUrn(itemUrn);
		resource.setType(req.getType());
		resource.setRawProperties(rawProps);
		resource.setProcessedProperties(processedProps);

		const item = new Item();
		item.setUrn(itemUrn);
		item.setMeta(req.getMeta()?.clone());
		item.setResource(resource);

		sessionRef.addItems(item);
		this._content.set(resource, content);

		return new v1_services_pb.CreateItemResponse().setItemUrn(item.getUrn());
	});
	updateItem = mockUnary<v1_services_pb.UpdateItemRequest, v1_services_pb.UpdateItemResponse>(async req => {
		await this.simulateNetwork();

		const itemRef = await this._getItemRef(req.getSessionUrn(), req.getItemUrn());
		itemRef.setMeta(req.getMeta()?.clone());
		return new v1_services_pb.UpdateItemResponse();
	});
	getItem = mockUnary<v1_services_pb.GetItemRequest, v1_services_pb.GetItemResponse>(async req => {
		await this.simulateNetwork();

		const res = new v1_services_pb.GetItemResponse();
		for (const urn of req.getItemUrnList()) {
			const itemRef = await this._getItemRef(req.getSessionUrn(), urn);
			res.addItems(itemRef.clone());
		}

		return res;
	});
	deleteItem = mockUnary<v1_services_pb.DeleteItemRequest, v1_services_pb.DeleteItemResponse>(async req => {
		await this.simulateNetwork();

		const sessionRef = await this._getSessionRef(req.getSessionUrn());
		const itemRef = await this._getItemRef(req.getSessionUrn(), req.getItemUrn());
		sessionRef.setItemsList(sessionRef.getItemsList().filter(i => i !== itemRef));

		return new v1_services_pb.DeleteItemResponse();
	});
}

function randomHex(length: number): string {
	return Array.from({ length }, () => Math.floor(Math.random() * 16).toString(16)).join("");
}
