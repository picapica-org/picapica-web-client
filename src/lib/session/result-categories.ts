import { Result } from "../generated/v1/types_pb";
import { DeepReadonly } from "../util";
import { PicapicaDocumentUrn, PicapicaItemUrn, PicapicaUrn } from "./urn";

export interface ItemResult {
	readonly a: PicapicaItemUrn;
	readonly b: PicapicaItemUrn;
	readonly result: DeepReadonly<Result.AsObject>;
}
export interface CollectionResult {
	readonly item: PicapicaItemUrn;
	readonly document: PicapicaDocumentUrn;
	readonly result: DeepReadonly<Result.AsObject>;
}

export interface ResultCategories {
	/**
	 * A list of all results between items.
	 */
	items: ItemResult[];
	/**
	 * A map from a collections URN to a non-empty list of results between an item and a document from key collection.
	 */
	collections: Map<string, CollectionResult[]>;
	/**
	 * A list of results that are invalid by not having valid URNs.
	 */
	invalid: DeepReadonly<Result.AsObject>[];
}

export function categorizeResults(results: Iterable<DeepReadonly<Result.AsObject>>): ResultCategories {
	const categories: ResultCategories = { items: [], collections: new Map(), invalid: [] };

	function addDocument(
		item: PicapicaItemUrn,
		document: PicapicaDocumentUrn,
		result: DeepReadonly<Result.AsObject>
	): void {
		const collectionUrn = PicapicaUrn.stringify({ type: "collection", collectionId: document.collectionId });

		let list = categories.collections.get(collectionUrn);
		if (list === undefined) {
			list = [];
			categories.collections.set(collectionUrn, list);
		}
		list.push({ item, document, result });
	}

	for (const result of results) {
		const a = PicapicaUrn.tryParse(result.resources?.urnA ?? "");
		const b = PicapicaUrn.tryParse(result.resources?.urnB ?? "");

		if (!a || !b) {
			categories.invalid.push(result);
			continue;
		}

		if (a.type === "item" && b.type === "item") {
			categories.items.push({ a, b, result });
		} else if (a.type === "item" && b.type === "document") {
			addDocument(a, b, result);
		} else if (a.type === "document" && b.type === "item") {
			addDocument(b, a, result);
		} else {
			categories.invalid.push(result);
		}
	}

	return categories;
}
