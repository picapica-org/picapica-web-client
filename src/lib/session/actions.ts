import { SessionConfig } from "../generated/v1/configs_pb";
import { DeleteItemRequest, Session, UpdateConfigRequest, UpdateItemRequest } from "../generated/v1/services_pb";
import { Item } from "../generated/v1/types_pb";
import { UseSessionArray } from "../use-session";
import { DeepReadonly } from "../util";
import { AnalysisConfig } from "./analysis-config";
import { cloneSession } from "./util";

export interface ActionResult<Req> {
	mutate: Parameters<UseSessionArray[1]>[1];
	request: Req;
}

export function deleteItemAction(
	session: DeepReadonly<Session.AsObject>,
	itemUrn: string
): ActionResult<DeleteItemRequest> {
	const request = new DeleteItemRequest();
	request.setSessionId(session.id);
	request.setItemId(itemUrn);

	return {
		request,
		mutate(oldSession) {
			const session = cloneSession(oldSession);
			session.itemsList = session.itemsList.filter(item => item.urn !== itemUrn);
			return session;
		},
	};
}

export function updateItemAction(
	session: DeepReadonly<Session.AsObject>,
	itemUrn: string,
	meta: Item.Metadata
): ActionResult<UpdateItemRequest> {
	const metaObject = meta.toObject();

	const request = new UpdateItemRequest();
	request.setSessionId(session.id);
	request.setItemId(itemUrn);
	request.setMeta(meta);

	return {
		request,
		mutate(oldSession) {
			const session = cloneSession(oldSession);
			const item = session.itemsList.find(item => item.urn === itemUrn);
			if (item) {
				item.meta = metaObject;
			}
			return session;
		},
	};
}

export function updateConfigAction(
	session: DeepReadonly<Session.AsObject>,
	analysisConfig: AnalysisConfig
): ActionResult<UpdateConfigRequest> {
	const pairs = analysisConfig.getResourcePairs();
	const pairingsList = pairs.map(p => p.toObject());

	const sessionConfig = new SessionConfig();
	sessionConfig.setPairingsList(pairs);

	const request = new UpdateConfigRequest();
	request.setSessionId(session.id);
	request.setConfig(sessionConfig);

	return {
		request,
		mutate(oldSession) {
			const session = cloneSession(oldSession);
			session.config ??= { pairingsList: [] };
			session.config.pairingsList = pairingsList;
			return session;
		},
	};
}
