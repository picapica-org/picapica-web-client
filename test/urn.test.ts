import { PicapicaUrn } from "../src/lib/session/urn";

test("parse", () => {
	function parse(urn: string): PicapicaUrn | Error {
		try {
			return PicapicaUrn.parse(urn);
		} catch (error) {
			if (error instanceof Error) {
				return error;
			}
			throw error;
		}
	}

	const urns = [
		// invalid
		"foo",
		"urn:foo:something",
		"urn:picapica:foo",
		"urn:picapica:item:324",

		// valid
		"urn:picapica:collection:dewiki",
		"urn:PicaPica:collection:dewiki",
		"urn:picapica:document:dewiki:e1bd8a08-64e3-4eed-a70d-5e435a88e707",
		"urn:picapica:session:f2270a54-88d6-4572-9b9c-9b60bfb43de3",
		"urn:picapica:item:f2270a54-88d6-4572-9b9c-9b60bfb43de3:a1244e8526a71095fe21825f073afe2a34f997e53874cd6986bdc049e218aced",
		"urn:picapica:result:f2270a54-88d6-4572-9b9c-9b60bfb43de3:a1244e8526a71095fe21825f073afe2a34f997e53874cd6986bdc049e218aced",
	];

	for (const urn of urns) {
		expect(parse(urn)).toMatchSnapshot();
	}
});
