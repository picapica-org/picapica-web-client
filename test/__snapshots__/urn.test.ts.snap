// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse 1`] = `[SyntaxError: Invalid Picapica URN: foo]`;

exports[`parse 2`] = `[SyntaxError: Invalid Picapica URN: urn:foo:something]`;

exports[`parse 3`] = `[SyntaxError: Invalid Picapica URN: urn:picapica:foo]`;

exports[`parse 4`] = `[SyntaxError: Invalid Picapica item URN: urn:picapica:item:324]`;

exports[`parse 5`] = `
{
  "collectionId": "dewiki",
  "type": "collection",
}
`;

exports[`parse 6`] = `
{
  "collectionId": "dewiki",
  "type": "collection",
}
`;

exports[`parse 7`] = `
{
  "collectionId": "dewiki",
  "documentId": "e1bd8a08-64e3-4eed-a70d-5e435a88e707",
  "type": "document",
}
`;

exports[`parse 8`] = `
{
  "sessionId": "f2270a54-88d6-4572-9b9c-9b60bfb43de3",
  "type": "session",
}
`;

exports[`parse 9`] = `
{
  "itemId": "a1244e8526a71095fe21825f073afe2a34f997e53874cd6986bdc049e218aced",
  "sessionId": "f2270a54-88d6-4572-9b9c-9b60bfb43de3",
  "type": "item",
}
`;

exports[`parse 10`] = `
{
  "resultId": "a1244e8526a71095fe21825f073afe2a34f997e53874cd6986bdc049e218aced",
  "sessionId": "f2270a54-88d6-4572-9b9c-9b60bfb43de3",
  "type": "result",
}
`;

exports[`parse 11`] = `
{
  "type": "none",
}
`;
